dadi/lib/providers/restapi.js

```diff
@@ -0,0 +1,240 @@
+'use strict'
+
+const url = require('url')
+const path = require('path')
+const request = require('request')
+const purest = require('purest')({ request })
+const purestConfig = require('@purest/providers')
+const DatasourceCache = require(path.join(__dirname, '/../cache/datasource'))
+
+const log = require('@dadi/logger')
+
+const help = require(path.join(__dirname, '../help'))
+
+const RestApi = function () {}
+
+/**
+ * initialise - initialises the datasource provider
+ *
+ * @param  {obj} datasource - the datasource to which this provider belongs
+ * @param  {obj} schema - the schema that this provider works with
+ * @return {void}
+ */
+RestApi.prototype.initialise = function initialise (datasource, schema) {
+  this.datasource = datasource
+  this.schema = schema
+
+  // If the provider is an object, pass the config to purest
+  // Else, use default @purest/providers
+  if (typeof this.datasource.source.provider === 'object') {
+    this.provider = Object.keys(this.datasource.source.provider)[0]
+    this.config = this.datasource.source.provider
+  } else {
+    this.provider = this.datasource.source.provider
+    this.config = purestConfig
+  }
+
+  this.purest = purest({
+    provider: this.provider,
+    config: this.config,
+    defaults: this.datasource.source.auth
+  })
+}
+
+/**
+ * load - loads data form the datasource
+ *
+ * @param  {string} requestUrl - url of the web request (not used)
+ * @param  {fn} done - callback on error or completion
+ * @return {void}
+ */
+RestApi.prototype.load = function load (requestUrl, done) {
+  log.info(
+    { module: 'restapi' },
+    'GET datasource "' + this.datasource.schema.datasource.key + '"'
+  )
+
+  // allow query string param to bypass cache
+  var query = url.parse(requestUrl, true).query
+  var noCache = query.cache && query.cache.toString().toLowerCase() === 'false'
+
+  try {
+    const endpoint = this.schema.datasource.source.endpoint
+    const query = this.query
+
+    this.cacheKey = [
+      endpoint,
+      encodeURIComponent(JSON.stringify(this.schema.datasource))
+    ].join('+')
+
+    this.dataCache = new DatasourceCache()
+
+    var cacheOptions = {
+      name: this.datasource.name,
+      caching: this.schema.datasource.caching,
+      cacheKey: this.cacheKey
+    }
+
+    this.dataCache.getFromCache(cacheOptions, cachedData => {
+      if (cachedData && !noCache) {
+        try {
+          cachedData = JSON.parse(cachedData.toString())
+          return done(null, cachedData)
+        } catch (err) {
+          console.error(
+            'Rest provider: cache data incomplete, making HTTP request: ' +
+              err +
+              '(' +
+              cacheOptions.cacheKey +
+              ')'
+          )
+        }
+      }
+
+      var self = this
```

Comment:
If we change the function on line 99 to be an arrow function, we could get rid of `self`?